nanofi/src/coap/c2protocol.c

```diff
@@ -0,0 +1,1204 @@
+                uint16_t num_params = cbor_get_uint16(num_params_item);
+                cbor_decref(&num_params_item);
+                ag_manifest->io.output_params[i].num_params = num_params;
+
+                ag_manifest->io.output_params[i].params = (char **)malloc(num_params * sizeof(char *));
+                int j;
+                for (j = 0; j < num_params; ++j) {
+                    cbor_item_t * param_item = cbor_load(payload + bytes_read, length - bytes_read, &result);
+                    if (result.error.code != CBOR_ERR_NONE || !cbor_isa_string(param_item)) {
+                        cbor_decref(&param_item);
+                        return -1;
+                    }
+                    bytes_read += result.read;
+                    size_t pl = cbor_string_length(param_item);
+                    ag_manifest->io.output_params[i].params[j] = (char *)malloc(pl + 1);
+                    memset(ag_manifest->io.output_params[i].params[j], 0, (pl + 1));
+                    const char * param_str = (const char *)cbor_string_handle(param_item);
+                    strncpy(ag_manifest->io.output_params[i].params[j], param_str, pl);
+                    cbor_decref(&param_item);
+                }
+            }
+        }
+    }
+    *bytes_decoded = bytes_read;
+    return 0;
+}
+
+int decode_heartbeat(const unsigned char * payload, size_t length, struct c2heartbeat * heartbeat, size_t * bytes_decoded) {
+    if (!heartbeat) {
+        return -1;
+    }
+    size_t bytes_read = 0;
+
+    struct cbor_load_result result;
+    cbor_item_t * cbor_item_type = cbor_load(payload + bytes_read, length - bytes_read, &result);
+    if (result.error.code != CBOR_ERR_NONE || !cbor_isa_uint(cbor_item_type)) {
+        cbor_decref(&cbor_item_type);
+        return -1;
+    }
+    bytes_read += result.read;
+    uint8_t is_heartbeat = cbor_get_uint8(cbor_item_type);
+    cbor_decref(&cbor_item_type);
+
+    if (!is_heartbeat) {
+        return -1;
+    }
+    size_t dev_info_bytes = 0;
+    if (decode_deviceinfo(payload + bytes_read, length - bytes_read, &heartbeat->device_info, &dev_info_bytes) < 0) {
+        free_c2heartbeat(heartbeat);
+        return -1;
+    }
+    bytes_read += dev_info_bytes;
+
+    size_t agent_info_bytes = 0;
+    if (decode_agentinfo(payload + bytes_read, length - bytes_read, &heartbeat->agent_info, &agent_info_bytes) < 0) {
+        free_c2heartbeat(heartbeat);
+        return -1;
+    }
+    bytes_read += agent_info_bytes;
+
+    size_t agent_manifest_bytes = 0;
+    if (decode_agentmanifest(payload + bytes_read, length - bytes_read, &heartbeat->ag_manifest, &agent_manifest_bytes) < 0) {
+        free_c2heartbeat(heartbeat);
+        return -1;
+    }
+    if (agent_manifest_bytes > 0) {
+        heartbeat->has_ag_manifest = 1;
+    }
+    bytes_read += agent_manifest_bytes;
+
+    *bytes_decoded = bytes_read;
+    return 0;
+}
+
+uint16_t endian_check_uint16(uint16_t value, int is_little_endian) {
```

Comment:
If we compile the code on a little endian machine and run the executable on big endian wouldn't that be problematic?

Reply:
Only the target platform matters. You can cross-compile e.g. from amd64 (little endian host platform) to arm (big endian target platform), in that case the generated binary will only work on arm machines and the compile-time check for endianness will return big endian, i.e. the endianness of the target platform rather than the host.

The hton* and ntoh* functions conveniently hide this and their API is flexible enough to even allow the hypothetical scenario of a CPU changing its endianness at runtime. The point is: the developer doesn't need to care how they work when correctly using the functions.